kedro/pipeline/pipeline.py

```diff
@@ -838,6 +934,179 @@ def to_json(self) -> str:
+
+        def _map_transcode_base(name: str) -> str:
+            base_name, transcode_suffix = _transcode_split(name)
+            return TRANSCODING_SEPARATOR.join((mapping[base_name], transcode_suffix))
+
+        rules = [
+            (lambda n: n in mapping, lambda n: mapping[n]),
+            (_is_all_parameters, lambda n: n),
+            (_is_transcode_base_in_mapping, _map_transcode_base),
+            (lambda n: bool(namespace) and _is_single_parameter(n), _prefix_param),
+            (lambda n: bool(namespace), _prefix_dataset),
+        ]
+
+        for predicate, processor in rules:
+            if predicate(name):  # type: ignore[no-untyped-call]
+                return processor(name)  # type: ignore[no-untyped-call, no-any-return]
+        return name
+
+    def _process_dataset_names(
+        self,
+        datasets: str | list[str] | dict[str, str] | None,
+        mapping: dict,
+        namespace: str | None,
+    ) -> str | list[str] | dict[str, str] | None:
+        if datasets is None:
+            return None
+        if isinstance(datasets, str):
+            return self._rename(datasets, mapping, namespace)
+        if isinstance(datasets, list):
+            return [self._rename(name, mapping, namespace) for name in datasets]
+        if isinstance(datasets, dict):
+            return {
+                key: self._rename(value, mapping, namespace)
+                for key, value in datasets.items()
+            }
+        raise ValueError(
+            f"Unexpected input {datasets} of type {type(datasets)}"
+        )  # pragma: no cover
+
+    def _copy_node(self, node: Node, mapping: dict, namespace: str | None) -> Node:
+        new_namespace = node.namespace
+        if namespace:
+            new_namespace = (
+                f"{namespace}.{node.namespace}" if node.namespace else namespace
+            )
+        return node._copy(
+            inputs=self._process_dataset_names(node._inputs, mapping, namespace),
+            outputs=self._process_dataset_names(node._outputs, mapping, namespace),
+            namespace=new_namespace,
+        )
+
+    def _map_nodes(
+        self,
+        pipe: Iterable[Node | Pipeline],
+        inputs: str | set[str] | dict[str, str] | None = None,
+        outputs: str | set[str] | dict[str, str] | None = None,
+        parameters: str | set[str] | dict[str, str] | None = None,
+        namespace: str | None = None,
+    ) -> list[Node]:
+        inputs = self._get_dataset_names_mapping(inputs)
+        outputs = self._get_dataset_names_mapping(outputs)
+        parameters = self._get_param_names_mapping(parameters)
+        _validate_datasets_exist(inputs.keys(), outputs.keys(), parameters.keys(), pipe)
+        _validate_inputs_outputs(inputs.keys(), outputs.keys(), pipe)
+        mapping = {**inputs, **outputs, **parameters}
+        new_nodes = [self._copy_node(n, mapping, namespace) for n in self._nodes]
+        return new_nodes
```

Comment:
I would've kept those as a carbon copy functions from the original, since all of them are private anyways and they have no need for `self` or the class. Btw as a follow up, we can potentially look into refactoring them through more concise naming or applying simpler operations somehow.